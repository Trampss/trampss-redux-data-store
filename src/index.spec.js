/* eslint-env jest */
import factory, { keyValue, simpleObject } from './index'

const Todo = id => ({ id, some: `information ${id}` })

const subState = {
  data: {
    1: Todo(1),
    20: Todo(20),
    2: Todo(2),
  },
  keys: [1, 20, 2],
  initialized: true,
}

describe('index', () => {
  const test = (reducer, state) => {
    it('should prefix actions', () => expect(reducer.add(Todo(2))).toMatchSnapshot())
    it('should prefix selectors', () => expect(reducer.get()(state)).toMatchSnapshot())
    it('should prefix reducer', () => expect(reducer(subState, reducer.remove(20))).toMatchSnapshot())
  }

  describe('with name and prefix', () => {
    test(
      factory({ key: 'id', name: 'todos', prefix: 'ui' }),
      { todos: subState },
    )
  })

  describe('with name and empty prefix', () => {
    test(
      factory({ key: 'id', name: 'todos', prefix: undefined }),
      { todos: subState },
    )
  })

  describe('with path', () => {
    test(
      factory({ key: 'id', path: 'api', name: 'todos' }),
      {
        api: {
          todos: subState,
        },
      },
    )
  })

  describe('without path', () => {
    test(factory({ key: 'id', name: 'todos' }), { todos: subState })
  })

  describe('with middleware', () => {
    const middleware = name => key => prefix => defaultData => ctx => ({
      ...ctx,
      state: { ...ctx.state, [name]: true, key, prefix, defaultData, prevCtx: ctx },
      action: { ...ctx.action, [name]: true },
    })

    test(
      factory({
        pre: [middleware('pre1')],
        post: [middleware('post1')],
      })({ key: 'id', name: 'todos' }),
      { todos: subState },
    )
  })

  describe('type validation', () => {
    it('should be on error [undefined]', () => {
      let error

      try {
        factory()
      } catch (ex) {
        error = ex
      }

      expect(error).toMatchSnapshot()
    })

    it('should be on error [not a middleware object, nor an option object]', () => {
      let error

      try {
        factory({ oups: 'unknown' })
      } catch (ex) {
        error = ex
      }

      expect(error).toMatchSnapshot()
    })

    it('should instanciate a reducer (options as object)', () => {
      expect(factory({ name: 'aa' }).krfType).toMatchSnapshot()
    })

    it('should instanciate a reducer (options as string -name-)', () => {
      expect(factory('a reducer').krfType).toMatchSnapshot()
    })
  })

  describe('named factory', () => {
    it('should instanciate a simpleObject reducer', () => {
      expect(simpleObject({ name: 'a simpleObject' }).krfType).toMatchSnapshot()
    })

    it('should instanciate a keyValue reducer (specified)', () => {
      expect(keyValue({ name: 'a keyValue' }).krfType).toMatchSnapshot()
    })

    it('should instanciate a keyValue reducer (default)', () => {
      expect(factory({ name: 'a keyValue' }).krfType).toMatchSnapshot()
    })
  })

  describe('extended selectors', () => {
    it('should extends isInitialized selector on simpleObject', () => {
      expect(simpleObject({ name: 'o' }).isInitialized({ o: '' })).toMatchSnapshot()
    })
    it('should extends get selector on simpleObject', () => {
      expect(simpleObject({ name: 'o', defaultData: {} }).get()({ o: 'DATA' })).toMatchSnapshot()
    })
  })

  describe('action type factories', () => {
    it('should export action type factories', () => {
      expect(simpleObject({ name: 'a simpleObject' }).SET).toMatchSnapshot()
    })
  })
})
