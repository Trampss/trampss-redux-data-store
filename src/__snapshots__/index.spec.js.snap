// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index action type factories should export action type factories 1`] = `"@@krf/SET_A SIMPLEOBJECT"`;

exports[`index extended selectors should extends get selector on simpleObject 1`] = `"DATA"`;

exports[`index extended selectors should extends isInitialized selector on simpleObject 1`] = `true`;

exports[`index named factory should instanciate a keyValue reducer (default) 1`] = `"keyValue"`;

exports[`index named factory should instanciate a keyValue reducer (specified) 1`] = `"keyValue"`;

exports[`index named factory should instanciate a simpleObject reducer 1`] = `"simpleObject"`;

exports[`index type validation should be on error [not a middleware object, nor an option object] 1`] = `[Error: parameter is not a middleware configuration, nor a factory option object.]`;

exports[`index type validation should be on error [undefined] 1`] = `[Error: parameter is not a middleware configuration, nor a factory option object.]`;

exports[`index type validation should instanciate a reducer (options as object) 1`] = `"keyValue"`;

exports[`index type validation should instanciate a reducer (options as string -name-) 1`] = `"keyValue"`;

exports[`index with middleware should prefix actions 1`] = `
Object {
  "payload": Object {
    "id": 2,
    "some": "information 2",
  },
  "type": "@@krf/ADD_TODOS",
}
`;

exports[`index with middleware should prefix reducer 1`] = `
Object {
  "array": Array [],
  "data": Object {
    "1": Object {
      "id": 1,
      "some": "information 1",
    },
    "2": Object {
      "id": 2,
      "some": "information 2",
    },
  },
  "defaultData": undefined,
  "initialized": true,
  "key": "id",
  "keys": Array [
    1,
    2,
  ],
  "post1": true,
  "pre1": true,
  "prefix": "todos",
  "prevCtx": Object {
    "action": Object {
      "payload": 20,
      "pre1": true,
      "type": "@@krf/REMOVE_TODOS",
    },
    "state": Object {
      "array": Array [],
      "data": Object {
        "1": Object {
          "id": 1,
          "some": "information 1",
        },
        "2": Object {
          "id": 2,
          "some": "information 2",
        },
      },
      "defaultData": undefined,
      "initialized": true,
      "key": "id",
      "keys": Array [
        1,
        2,
      ],
      "pre1": true,
      "prefix": "todos",
      "prevCtx": Object {
        "action": Object {
          "payload": 20,
          "type": "@@krf/REMOVE_TODOS",
        },
        "state": Object {
          "data": Object {
            "1": Object {
              "id": 1,
              "some": "information 1",
            },
            "2": Object {
              "id": 2,
              "some": "information 2",
            },
            "20": Object {
              "id": 20,
              "some": "information 20",
            },
          },
          "initialized": true,
          "keys": Array [
            1,
            20,
            2,
          ],
        },
      },
    },
  },
}
`;

exports[`index with middleware should prefix selectors 1`] = `
Object {
  "1": Object {
    "id": 1,
    "some": "information 1",
  },
  "2": Object {
    "id": 2,
    "some": "information 2",
  },
  "20": Object {
    "id": 20,
    "some": "information 20",
  },
}
`;

exports[`index with name and empty prefix should prefix actions 1`] = `
Object {
  "payload": Object {
    "id": 2,
    "some": "information 2",
  },
  "type": "@@krf/ADD_TODOS",
}
`;

exports[`index with name and empty prefix should prefix reducer 1`] = `
Object {
  "array": Array [],
  "data": Object {
    "1": Object {
      "id": 1,
      "some": "information 1",
    },
    "2": Object {
      "id": 2,
      "some": "information 2",
    },
  },
  "initialized": true,
  "keys": Array [
    1,
    2,
  ],
}
`;

exports[`index with name and empty prefix should prefix selectors 1`] = `
Object {
  "1": Object {
    "id": 1,
    "some": "information 1",
  },
  "2": Object {
    "id": 2,
    "some": "information 2",
  },
  "20": Object {
    "id": 20,
    "some": "information 20",
  },
}
`;

exports[`index with name and prefix should prefix actions 1`] = `
Object {
  "payload": Object {
    "id": 2,
    "some": "information 2",
  },
  "type": "@@krf/ADD_UITODOS",
}
`;

exports[`index with name and prefix should prefix reducer 1`] = `
Object {
  "array": Array [],
  "data": Object {
    "1": Object {
      "id": 1,
      "some": "information 1",
    },
    "2": Object {
      "id": 2,
      "some": "information 2",
    },
  },
  "initialized": true,
  "keys": Array [
    1,
    2,
  ],
}
`;

exports[`index with name and prefix should prefix selectors 1`] = `
Object {
  "1": Object {
    "id": 1,
    "some": "information 1",
  },
  "2": Object {
    "id": 2,
    "some": "information 2",
  },
  "20": Object {
    "id": 20,
    "some": "information 20",
  },
}
`;

exports[`index with path should prefix actions 1`] = `
Object {
  "payload": Object {
    "id": 2,
    "some": "information 2",
  },
  "type": "@@krf/ADD_TODOS",
}
`;

exports[`index with path should prefix reducer 1`] = `
Object {
  "array": Array [],
  "data": Object {
    "1": Object {
      "id": 1,
      "some": "information 1",
    },
    "2": Object {
      "id": 2,
      "some": "information 2",
    },
  },
  "initialized": true,
  "keys": Array [
    1,
    2,
  ],
}
`;

exports[`index with path should prefix selectors 1`] = `
Object {
  "1": Object {
    "id": 1,
    "some": "information 1",
  },
  "2": Object {
    "id": 2,
    "some": "information 2",
  },
  "20": Object {
    "id": 20,
    "some": "information 20",
  },
}
`;

exports[`index without path should prefix actions 1`] = `
Object {
  "payload": Object {
    "id": 2,
    "some": "information 2",
  },
  "type": "@@krf/ADD_TODOS",
}
`;

exports[`index without path should prefix reducer 1`] = `
Object {
  "array": Array [],
  "data": Object {
    "1": Object {
      "id": 1,
      "some": "information 1",
    },
    "2": Object {
      "id": 2,
      "some": "information 2",
    },
  },
  "initialized": true,
  "keys": Array [
    1,
    2,
  ],
}
`;

exports[`index without path should prefix selectors 1`] = `
Object {
  "1": Object {
    "id": 1,
    "some": "information 1",
  },
  "2": Object {
    "id": 2,
    "some": "information 2",
  },
  "20": Object {
    "id": 20,
    "some": "information 20",
  },
}
`;
